File: Scanner/src/parser/Parser.java
package parser;
import scanner.Scanner;
import environment.Environment;
import ast.*;
import scanner.ScanErrorException;
import java.io.*;
import java.util.List;
import java.util.ArrayList;

/**
 * Parser uses the Scanner to read a file and process a language.
 * 
 * @author David Melisso
 * @version February 28, 2018
 *
 */
public class Parser
{
    private Scanner scanner;
    private String currentToken;
    private Environment global;
    private java.util.Scanner read;
   
   /**
    * Constructs a Parser using a Scanner as input and sets the first
    * currentToken using the first token from said Scanner.
    * 
    * @param s
    *       the scanner to input from
    */
    public Parser(Scanner s) 
    {
        scanner = s;
        read = new java.util.Scanner(System.in);
        global = new Environment(read);
        try 
        {
            currentToken = scanner.nextToken();
        }
        catch (ScanErrorException e)
        {
            e.printStackTrace();
            System.exit(1);
        }
    }
   
   /**
    * Method: hasNext
    * 
    * Determines whether the scanner has another token
    * 
    * @return false if the input stream has ended; otherwise, true
    */
    public boolean hasNext()
    {
        return scanner.hasNext();
    }
   
   /**
    * Checks if currentToken matches the expected character;
    * if it does, moves the scanner to the next token,
    * otherwise, throws an IllegalArgumentException
    * 
    * @param expected
    *       the toke  to expect
    * @throws IllegalArgumentException
    *   if the expected token doesn't match the current token
    */
    private void eat(String expected) throws IllegalArgumentException
    {
        if (currentToken.equals(expected))
        {
            try
            {
                currentToken = scanner.nextToken();
            }
            catch (ScanErrorException e)
            {
                e.printStackTrace();
                System.exit(1);
            }
        }
        else
        {
            throw new IllegalArgumentException(
                   "Unexpected token " + currentToken + 
                   ", expected " + expected);
           
        }
    }
   
   /**
    * Scans the current token as an integer and returns it
    * 
    * @precondition current token is an integer
    * @postcondition number token has been eaten
    * @return the value of the parsed integer
    * @throws NumberFormatException
    *   if the token is not an integer
    */
    private ast.Number parseNumber() throws NumberFormatException
    {
        int res = Integer.parseInt(currentToken);
        eat(currentToken);
        return new ast.Number(res);
    }
    
    private List<Expression> eatParams()
    {
        List<Expression> params = new ArrayList<Expression>();
        eat("(");
        
        while (!currentToken.equals(")"))
        {
            params.add(parseExpr());
            
            while (currentToken.equals(","))
            {
                eat(",");
                params.add(parseExpr());
            }
        }
        
        eat(")");
        
        return params;
    }
   
   /**
    * Parses the statement:
    *   BEGIN ... END; where the program parses all the statements in between
    *   WRITELN(num); where the program prints the integer num
    *   READLN(var); where the program reads user input and sets it to 
    *       the variable var
    *   IF cond THEN stmt; where the program executes stmt if cond is true
    *   IF cond THEN stmt1 ELSE stmt2; where the program executes stmt1 if cond is true,
    *       and executes stmt2 if cond is false
    *   WHILE cond DO stmt; where the program executes stmt while cond is true
    *   FOR assignment TO number DO stmt; where the program increments the variable in 
    *       the assignment towards that number and executes stmt every interval
    *   BREAK; where the program exits the current loop
    *   CONTINUE; where the program skips the current iteration of the loop
    *   FUNCTION func(params...): stmt; where the program defins a function named func
    *       with parameters params which runs the statement stmt
    *   func(params...); which calls the function func with given parameters (therefore
    *       in a new environment)
    *   var := num; where the program sets the variable var to num
    * 
    * @precondition currentToken is WRITELN
    * @return the Statement of the correct type
    */
    public Statement parseStatement()
    {
        // BEGIN block
        if (currentToken.equals("BEGIN"))
        {
            List<Statement> stmts = new ArrayList<Statement>();
            eat("BEGIN");
            while (!currentToken.equals("END"))
            {
                stmts.add(parseStatement());
            }
            eat("END");
            eat(";");
            return new Block(stmts);
        }
        // IF statement
        if (currentToken.equals("IF"))
        {
            eat("IF");
            Condition cond = parseBool();
            eat("THEN");
            Statement stmt = parseStatement();
            if (currentToken.equals("ELSE"))
            {
                eat("ELSE");
                return new If(cond, stmt, parseStatement());
            }
            return new If(cond, stmt);
        }
        // WHILE loop statement
        if (currentToken.equals("WHILE"))
        {
            eat("WHILE");
            Condition cond = parseBool();
            eat("DO");
            return new While(cond, parseStatement());
        }
        // FOR loop statement
        if (currentToken.equals("FOR"))
        {
            eat("FOR");
            String var = currentToken;
            eat(currentToken);
            eat(":=");
            Expression expr = parseExpr();
            eat("TO");
            Expression limit = parseExpr();
            eat("DO");
            Statement stmt = parseStatement();
            
            Variable variable = new Variable(var);
            Assignment initialAssignment = new Assignment(var, expr);
            Expression increment = new BinOp(variable, "+", new ast.Number(1));
            Statement incrementAssignment = new Assignment(var, increment);
            
            // create loop sub-statement and increment
            List<Statement> loopComponents = new ArrayList<Statement>();
            loopComponents.add(stmt);
            loopComponents.add(incrementAssignment);
            Statement loopStatement = new Block(loopComponents);
            
            // create loop
            Condition loopCondition = new Condition(variable,"<=",limit);
            Statement loop = new While(loopCondition, loopStatement);
            
            // create loop and initial assignment
            List<Statement> assignmentAndLoop = new ArrayList<Statement>();
            assignmentAndLoop.add(initialAssignment);
            assignmentAndLoop.add(loop);
            
            return new Block(assignmentAndLoop);
        }
        // BREAK statement
        if (currentToken.equals("BREAK"))
        {
            eat("BREAK");
            eat(";");
            return new Break();
        }
        // CONTINUE statement
        if (currentToken.equals("CONTINUE"))
        {
            eat("CONTINUE");
            eat(";");
            return new Continue();
        }
        // RETURN statement
        if (currentToken.equals("RETURN"))
        {
            eat("RETURN");
            if (currentToken.equals(";"))
            {
                eat(";");
                return new Return();
            }
            Return res = new Return(parseExpr());
            eat(";");
            return res;
        }
        // WRITELN statement
        if (currentToken.equals("WRITELN"))
        {
            eat("WRITELN");
            eat("(");
            //System.out.println(parseStrings());
            Expression expr = parseStrings();//parseExpr();
            eat(")");
            eat(";");
            return new Writeln(expr);
        }
        // READLN statement
        if (currentToken.equals("READLN"))
        {
            eat("READLN");
            eat("(");
            String var = currentToken;
            eat(currentToken);
            eat(")");
            eat(";");
            return new Readln(var);
        }
        // FUNCTION statement
        if (currentToken.equals("FUNCTION"))
        {
            eat("FUNCTION");
            
            if (!Scanner.isLetter(currentToken.charAt((0))))
            {
                throw new Error("Invalid function identifier " + currentToken);
            }
            String funcName = currentToken;
            eat(funcName);
            
            List<String> params = new ArrayList<String>();
            eat("(");
            
            while (!currentToken.equals(")"))
            {
                params.add(currentToken);
                eat(currentToken);
                
                while (currentToken.equals(","))
                {
                    eat(",");
                    params.add(currentToken);
                    eat(currentToken);
                }
            }
            
            eat(")");
            
            
            eat(":");
            Statement s = parseStatement();
            return new FunctionAssignment(funcName, params, s);
        }
        // ensure identifier
        if (!Scanner.isLetter(currentToken.charAt((0))))
        {
            System.err.println("Error: invalid identifier " + currentToken);
            System.exit(1);
            return null;
        }
        // set variable or call function
        String var = currentToken;
        eat(currentToken);
        
        // set variable
        if (currentToken.equals(":="))
        {
            eat(":=");
            Expression expr = parseExpr();
            eat(";");
            return new Assignment(var, expr);
        }
        
        // call function
        if (currentToken.equals("("))
        {
            List<Expression> params = eatParams();
            FunctionCallStatement res = new FunctionCallStatement(var, params);
            eat(";");
            return res;
        }
        
        throw new IllegalArgumentException("Expected \":=\" or \"(\", read " + 
                            currentToken);
    }
   
   /**
    * Parses one or more strings, connected by a comma
    * 
    * @return a single string, concatenating the multiple strings
    */
    public Text parseStrings()
    {
        List<Text> current = new ArrayList<Text>();
        current.add(parseString());
        while (currentToken.charAt(0) == ',')
        {
            eat(",");
            current.add(parseString());
        }
        return new Text(current);
    }
   
   /**
    * Parses a single string by removing the surrounding quotes, or an expression
    * as a string.
    * 
    * @return the parsed text
    */
    public Text parseString()
    {
        if (currentToken.charAt(0) == '\'')
        {
            Text res = new Text(currentToken.substring(1, currentToken.length() - 1));
            eat(currentToken);
            return res;
        }
        return new Text(parseExpr());
    }
   
    /**
     * Parses a boolean (condition), which can be:
     * 
     * expr; where the expression is evaluated as a boolean
     * expr op expr; where the expression is compared to another expression with a
     *      comparison operator
     * 
     * @return the Condition which represents the boolean
     */
    private Condition parseBool()
    {
       /*
       if (currentToken.equals("TRUE"))
       {
           eat("TRUE");
           return true;
       }
       if (currentToken.equals("FALSE"))
       {
           eat("FALSE");
           return false;
       }
       if (currentToken.equals("NOT"))
       {
           eat("NOT");
           return !parseBool();
       }
       */
       
        Expression expr1 = parseExpr();
           
        if (currentToken.equals(">") ||
                currentToken.equals("<") ||
                currentToken.equals(">=") ||
                currentToken.equals("<=") ||
                currentToken.equals("<>") ||
                currentToken.equals("="))
           
        {
            String op = currentToken;
            eat(op);
            return new Condition(expr1, op, parseExpr());
        }
        return new Condition(expr1);
    }
   
   /**
    * Parses a factor, which can be:
    *   - factor        => returns -1 times the next factor
    *   ( expr )        => returns the expr within the ( )
    *   id              => returns the value of the variable
    *   func()          => returns the result of the function
    *   num             => returns the value of the number
    * where num is a number and id is an identifier.
    * 
    * @return the parsed factor
    */
    private Expression parseFactor()
    {
        // - factor
        if (currentToken.equals("-"))
        {
            eat("-");
            return new BinOp(new ast.Number(0), "-", parseFactor());
        }
        // ( expr )
        else if (currentToken.equals("("))
        {
            eat("(");
            Expression res = parseExpr();
            //System.out.println(res);
            eat(")");
            return res;
        }
       /*else if (currentToken.equals(","))
       {
           eat(",");
           return null;
       }*/
       // identifier (variable or function)
        else if (Scanner.isLetter(currentToken.charAt(0)))
        {
            String id = currentToken;
            eat(id);
            
            // function
            if (currentToken.equals("("))
            {
                return new FunctionCall(id, eatParams());
            }
            
            // variable
            return new Variable(id);
        }
       // number or string
        else
        {
            try
            {
                return parseNumber();
            }
            catch(NumberFormatException e)
            {
                return parseStrings();
            }
        }
    }
   
   /**
    * Parses an expr, which can be:
    *   expr (+ | -) term       => returns either the sum or difference
    *                              of the two values
    *   term                    => returns the value of the term
    *   string                  => returns the string
    *   
    *   
    * @return the parsed expr
    */
    private Expression parseExpr()
    {
        Expression cur = parseTerm();
        if (cur instanceof Text)
        {
            if (currentToken.equals("*") || 
                    currentToken.equals("/") ||
                    currentToken.equals("mod"))
            {
                try
                {
                    cur = ((Text) cur).toNum();
                }
                catch(NumberFormatException e)
                {
                    throw new Error("Cannot use operator " + currentToken +
                            " between a String and an Integer");
                }
            }
            return cur;
        }
       
        while (currentToken.equals("+") || currentToken.equals("-"))
        {
            String operator = currentToken;
            eat(currentToken);
            cur = new BinOp(cur, operator, parseTerm());
        }
        return cur;
    }
   
   /**
    * Parses an term, which can be:
    *   term (* | / | mod) factor   => returns the result of the
    *                                  evaluated term
    *   factor                      => returns the value of the factor
    *   string                      => returns the string
    *   
    * @return the parsed term
    */
    private Expression parseTerm()
    {
        Expression cur = parseFactor();
        if (cur instanceof Text)
        {
            if (currentToken.equals("*") || 
                    currentToken.equals("/") ||
                    currentToken.equals("mod"))
            {
                try
                {
                    cur = ((Text) cur).toNum();
                }
                catch(NumberFormatException e)
                {
                    throw new Error("Cannot use operator " + currentToken +
                            " between a String and an Integer");
                }
            }
            return cur;
        }
       
        while (currentToken.equals("*") || 
                currentToken.equals("/") ||
                currentToken.equals("mod"))
        {
            String operator = currentToken;
            eat(currentToken);
            cur = new BinOp(cur, operator, parseFactor());
        }
        return cur;
    }
    
    /**
     * Parses all the statements in the code in a given environment.
     * @param env
     *  the environment to execute in
     * @throws ASTException
     *  if an error occurs while parsing the statement
     */
    public void parseAndExecute(Environment env) throws ASTException
    {
        while (hasNext())
        {
            parseStatement().exec(env);
        }
    }
    
    /**
     * Parses all the statements in the code in the global environment.
     * @throws ASTException
     *  if an error occurs while parsing the statement
     */
    public void parseAndExecute() throws ASTException
    {
        while (hasNext())
        {
            parseStatement().exec(global);
        }
    }
   
   
   /**
    * Parses all the statements on file ./HomemadeParserTest.txt
    * 
    * @param args
    *   command line arguments
    */
    public static void main(String[] args)
    {
        String filename =  "./HomemadeParserTest.txt";
        try
        {
            Parser p = (new Parser(
                    new Scanner(
                           new FileInputStream(
                                   new File(filename)))));
           /*while(p.hasNext())
           {
               p.parseStatement();
           }*/
            p.parseAndExecute();
        }
        catch (FileNotFoundException | ASTException e)
        {
            e.printStackTrace();
        }
    }
}

File: JavaScriptInterpreter/parser/parser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var scanner_1 = __importDefault(require("../scanner/scanner"));
var boolean_1 = __importDefault(require("../ast/boolean"));
var number_1 = __importDefault(require("../ast/number"));
var text_1 = __importDefault(require("../ast/text"));
var variable_1 = __importDefault(require("../ast/variable"));
var binop_1 = __importDefault(require("../ast/binop"));
var display_1 = __importDefault(require("../ast/display"));
var assignment_1 = __importDefault(require("../ast/assignment"));
var while_1 = __importDefault(require("../ast/while"));
var if_1 = __importDefault(require("../ast/if"));
var program_1 = __importDefault(require("../ast/program"));
var environment_1 = __importDefault(require("../environment/environment"));
/**
 * Parser parses an input from a Scanner and returns an Abstract Syntax Tree through
 * its method parseProgram.
 */
var Parser = /** @class */ (function () {
    /**
     * Creates a new Parser using a Scanner, and gets the first Token
     * @param {Scanner | string} inStream the Scanner or string which is fed into the parser
     */
    function Parser(inStream) {
        if (inStream instanceof scanner_1.default) {
            this.scanner = inStream;
        }
        else {
            this.scanner = new scanner_1.default(inStream);
        }
        this.currentToken = this.scanner.nextToken();
    }
    /**
     * Compares a string to the current token and throws an error if they are different
     * @param {string} expected the string to compare to the current token
     */
    Parser.prototype.eat = function (expected) {
        if (this.currentToken == expected) {
            this.currentToken = this.scanner.nextToken();
        }
        else {
            throw new Error("[ERROR] Unexpected token " + this.currentToken +
                ", expected " + expected);
        }
    };
    /**
     * Determines whether a string is an identifier
     * An identifier is any string which starts with a letter and is followed by numbers or letters
     * @param {string} str the string to evaluate as an identifier
     * @returns true if str is an identifier;
     *          false otherwise
     */
    Parser.isIdentifier = function (str) {
        return scanner_1.default.isLetter(str.charAt(0));
    };
    /**
     * Reads in a *Value* from the input scanner
     *
     * *Value*s can be of the form:
     * - `(`*Expression*`)`
     * - `true`
     * - `false`
     * - *id*
     * - *num*
     *
     * @returns the Expression correlating to the *Value* found
     */
    Parser.prototype.parseValue = function () {
        // (Expression)
        if (this.currentToken.charAt(0) == "(") {
            this.eat("(");
            var temp = this.parseExpression();
            this.eat(")");
            return temp;
        }
        // boolean
        if (this.currentToken == "true") {
            this.eat("true");
            return new boolean_1.default(true);
        }
        if (this.currentToken == "false") {
            this.eat("false");
            return new boolean_1.default(false);
        }
        // id
        if (Parser.isIdentifier(this.currentToken)) {
            var res_1 = this.currentToken;
            this.eat(res_1);
            return new variable_1.default(res_1);
        }
        // number
        var res = parseInt(this.currentToken);
        this.eat(this.currentToken);
        return new number_1.default(res);
    };
    /**
     * Reads in a *NegExpr* from the input scanner
     *
     * *NegExpr*s can be of the form:
     * - *Value*
     * - `-`*Value*
     *
     * @returns the Expression correlating to the *NegExpr* found
     */
    Parser.prototype.parseNegExpr = function () {
        if (this.currentToken.charAt(0) == "-") {
            this.eat("-");
            return new binop_1.default(new number_1.default(0), "-", this.parseValue());
        }
        return this.parseValue();
    };
    /**
     * Reads in a *MultExpr* from the input scanner
     *
     * *MultExpr*s can be of the form:
     * - *NegExpr*
     * - *NegExpr* `*` *NegExpr*
     * - *NegExpr* `/` *NegExpr*
     *
     * @returns the Expression correlating to the *MultExpr* found
     */
    Parser.prototype.parseMultExpr = function () {
        var res = this.parseNegExpr();
        while (this.currentToken == "*" || this.currentToken == "/") {
            var operator = this.currentToken;
            this.eat(operator);
            var nextTerm = this.parseNegExpr();
            res = new binop_1.default(res, operator, nextTerm);
        }
        return res;
    };
    /**
     * Reads in a *AddExpr* from the input scanner
     *
     * *AddExpr*s can be of the form:
     * - *MultExpr*
     * - *MultExpr* `+` *MultExpr*
     * - *MultExpr* `-` *MultExpr*
     *
     * @returns the Expression correlating to the *AddExpr* found
     */
    Parser.prototype.parseAddExpr = function () {
        var res = this.parseMultExpr();
        while (this.currentToken == "+" || this.currentToken == "-") {
            var operator = this.currentToken;
            this.eat(operator);
            var nextTerm = this.parseMultExpr();
            res = new binop_1.default(res, operator, nextTerm);
        }
        return res;
    };
    /**
     * Reads in a *Expression* from the input scanner
     *
     * *AddExpr*s can be of the form:
     * - *AddExpr*
     * - *AddExpr* `<` *AddExpr*
     * - *AddExpr* `>` *AddExpr*
     * - *AddExpr* `>=` *AddExpr*
     * - *AddExpr* `<=` *AddExpr*
     * - *AddExpr* `<>` *AddExpr*
     * - *AddExpr* `=` *AddExpr*
     *
     * @returns the Expression correlating to the *Expression* found
     */
    Parser.prototype.parseExpression = function () {
        var res = this.parseAddExpr();
        while (this.currentToken == "<"
            || this.currentToken == ">"
            || this.currentToken == ">="
            || this.currentToken == "<="
            || this.currentToken == "<>"
            || this.currentToken == "=") {
            var operator = this.currentToken;
            this.eat(operator);
            var nextTerm = this.parseAddExpr();
            res = new binop_1.default(res, operator, nextTerm);
        }
        return res;
    };
    /**
     * Reads in a *Display* Statement from the Scanner
     * Display Statements come both of these forms:
     *
     * - `display` *Expression*
     * - `display` *Expression* `read` *id*
     *
     * Both print the evaluated *Expression* to the console; but
     * the second one reads a value from the user and sets it to
     * a variable with the given identifier
     *
     * @returns a Display Statement correlating to the input from the Scanner
     */
    Parser.prototype.parseDisplay = function () {
        this.eat("display");
        var expr;
        if (this.currentToken.charAt(0) == "'") {
            expr = new text_1.default(this.currentToken.substr(1, this.currentToken.length - 2));
            this.eat(this.currentToken);
        }
        else {
            expr = this.parseExpression();
        }
        if (this.currentToken == "read") {
            this.eat("read");
            var saveTo = this.currentToken;
            this.eat(saveTo);
            return new display_1.default(expr, saveTo);
        }
        return new display_1.default(expr);
    };
    /**
     * Reads in a *Assignment* Statement from the Scanner
     * Assignment Statements come both of these forms:
     *
     * - `assign` *id* `=` *Expression*
     *
     * This assigns the evaluated *Expression* to the variable with the given *id*
     *
     * @returns a Assignment Statement correlating to the input from the Scanner
     */
    Parser.prototype.parseAssign = function () {
        this.eat("assign");
        var toVar = this.currentToken;
        this.eat(toVar);
        this.eat("=");
        return new assignment_1.default(toVar, this.parseExpression());
    };
    /**
     * Reads in a *While* Statement from the Scanner
     * While Statements come both of these forms:
     *
     * - `while` *Expression* `do` *Program* `end`
     *
     * This evaluates the *Program* every time the *Expression* evaluates to true
     *
     * @returns a While Statement correlating to the input from the Scanner
     */
    Parser.prototype.parseWhile = function () {
        this.eat("while");
        var condition = this.parseExpression();
        this.eat("do");
        var prog = this.parseProgram();
        this.eat("end");
        return new while_1.default(condition, prog);
    };
    /**
     * Reads in a *If* Statement from the Scanner
     * If Statements come both of these forms:
     *
     * - `if` *Expression* `then` *Program* `end`
     * - `if` *Expression* `then` *Program* `else` *Program* `end`
     *
     * If the expression evaluates to `true`, then the first *Program* is evaluated;
     * If the expression evaluates to `false` and there is a second *Program*, then it runs
     * the second *Program*
     *
     * @returns an If Statement correlating to the input from the Scanner
     */
    Parser.prototype.parseIf = function () {
        this.eat("if");
        var cond = this.parseExpression();
        this.eat("then");
        var then = this.parseProgram();
        if (this.currentToken == "else") {
            this.eat("else");
            var els = this.parseProgram();
            this.eat("end");
            return new if_1.default(cond, then, els);
        }
        this.eat("end");
        return new if_1.default(cond, then);
    };
    /**
     * Parses a statement depending on the current Token
     *
     * If the current token is `display`, it parses a Display Statement
     * If the current token is `assign`, it parses a Assignment Statement
     * If the current token is `while`, it parses a While Statement
     * If the current token is `if`, it parses an If Statement
     *
     * @returns the Statement correlating to the given input from the Scanner
     * @throws Error if the current token is not `display`, `assign`, `while`, or `if`
     */
    Parser.prototype.parseStatement = function () {
        if (this.currentToken == "display")
            return this.parseDisplay();
        if (this.currentToken == "assign")
            return this.parseAssign();
        if (this.currentToken == "while")
            return this.parseWhile();
        if (this.currentToken == "if")
            return this.parseIf();
        throw new Error("invalid statement: " + this.currentToken);
    };
    /**
     * Parses a Program, which is an array of Statements, until it reaches an endpoint
     * An enpoint occurs when it the code ends, or it reaches the token `end` or `else`
     *
     * @returns the Program containing all of the Statements
     */
    Parser.prototype.parseProgram = function () {
        var res = [];
        res.push(this.parseStatement());
        while (this.scanner.hasNext() && this.currentToken != "end" && this.currentToken != "else") {
            res.push(this.parseStatement());
        }
        return new program_1.default(res);
    };
    /**
     * Runs a test of the Parser, parsing the file `ParserTest.txt` in this directrory
     */
    Parser.test = function () {
        var filename = __dirname + "/ParserTest.txt";
        try {
            var p = new Parser(filename);
            p.parseProgram().exec(new environment_1.default());
        }
        catch (e) {
            console.error(e);
            process.exit(1);
        }
    };
    return Parser;
}());
exports.default = Parser;
// If this is file is being run, run the test
if (require.main === module) {
    Parser.test();
}
